gax/src/main/java/com/google/api/gax/rpc/StubSettings.java

```diff
@@ -70,11 +70,20 @@
   private final TransportChannelProvider transportChannelProvider;
   private final ApiClock clock;
   private final String endpoint;
+  private final String mtlsEndpoint;
   private final String quotaProjectId;
   @Nullable private final WatchdogProvider streamWatchdogProvider;
   @Nonnull private final Duration streamWatchdogCheckInterval;
   @Nonnull private final ApiTracerFactory tracerFactory;
 
+  /**
+   * Indicate when creating transport whether it is allowed to use mTLS endpoint instead of the
+   * default endpoint. Only the endpoint set by client libraries is allowed. User provided endpoint
+   * should always be used as it is. Client libraries can set it via {@link
+   * Builder#setSwitchToMtlsEndpointAllowed} method.
+   */
+  private final boolean switchToMtlsEndpointAllowed;
```

Comment:
What's the point of having an MTLS endpoint if it can't be switched to? 

If mtlsEndpoint is always null when switchToMtlsEndpointAllowed is false, then perhaps you don't need switchToMtlsEndpointAllowed.

Reply:
Client libraries subclass StubSettings, for instance KmsStubSettings for cloud KMS. In the subclass, mtls endpoint will always be set to the default mtls endpoint, which is never null, regardless of the `switchToMtlsEndpointAllowed` value. The `switchToMtlsEndpointAllowed` value depends on how the users build the KmsStubSettings instance. 
(1) If the user calls `setEndpoint` in `KmsStubSettings.Builder`, then `switchToMtlsEndpointAllowed` is false and we will use whatever endpoint the user provides (in other words, we are not allowed to switch endpoints)
(2) If the user doesn't call `setEndpoint`, then `switchToMtlsEndpointAllowed` is true and switch from the regular endpoint to mtls endpoint is allowed.